File: Sources/SSHTunnelManager/Models/SSHProfile.swift
import Foundation

/// Shared limits for the terminal text size (points). Used by profiles, the
/// local-terminal default, and the ⌘+/⌘− zoom commands.
enum TerminalFontMetrics {
    static let `default`: Double = 13
    static let min: Double = 8
    static let max: Double = 36
    static let step: Double = 1

    /// Keep a size within the allowed range (and rounded to a whole point).
    static func clamp(_ size: Double) -> Double {
        Swift.min(max, Swift.max(min, (size).rounded()))
    }
}

/// The kind of SSH port forwarding.
enum ForwardType: String, Codable, CaseIterable, Identifiable {
    case local
    case remote
    case dynamic

    var id: String { rawValue }

    /// The `ssh` command-line flag for this forward type.
    var flag: String {
        switch self {
        case .local: return "-L"
        case .remote: return "-R"
        case .dynamic: return "-D"
        }
    }

    var title: String {
        switch self {
        case .local: return "Local  ·  -L"
        case .remote: return "Remote  ·  -R"
        case .dynamic: return "Dynamic / SOCKS  ·  -D"
        }
    }

    var explanation: String {
        switch self {
        case .local:
            return "Opens a port on THIS Mac and forwards it through the server to a target reachable from the server. (e.g. reach a remote database locally)"
        case .remote:
            return "Opens a port on the SERVER and forwards it back to a target reachable from this Mac. (e.g. expose a local service to the server)"
        case .dynamic:
            return "Runs a SOCKS proxy on this Mac; apps pointed at it route their traffic through the server."
        }
    }
}

/// What a (local) port forward exposes, so the app can offer a matching
/// "Open" action that launches the right kind of tab against the forwarded
/// local port. Purely a convenience layer over a normal `-L` forward — it does
/// not change the `ssh` command at all.
enum ForwardCategory: String, Codable, CaseIterable, Identifiable {
    case none
    case webpage
    case mqtt
    case redis

    var id: String { rawValue }

    /// Menu / picker label.
    var title: String {
        switch self {
        case .none:    return "None"
        case .webpage: return "Web Page"
        case .mqtt:    return "MQTT"
        case .redis:   return "Redis"
        }
    }

    /// SF Symbol used in pickers, launchers and (for mqtt/redis) the tab itself.
    var symbol: String {
        switch self {
        case .none:    return "minus.circle"
        case .webpage: return "globe"
        case .mqtt:    return "antenna.radiowaves.left.and.right"
        case .redis:   return "cylinder.split.1x2"
        }
    }

    /// Whether this category can launch a tab (everything but `.none`).
    var isLaunchable: Bool { self != .none }

    /// The conventional default port for this service, used to prefill the
    /// ad-hoc “new connection” form.
    var defaultPort: Int {
        switch self {
        case .webpage: return 8080
        case .mqtt:    return 1883
        case .redis:   return 6379
        case .none:    return 0
        }
    }

    /// The terminal-tab kind a launchable service opens, when it's a CLI client
    /// (mqtt / redis). `.webpage` opens a browser tab instead, so it's nil here.
    var terminalKind: TerminalSession.Kind? {
        switch self {
        case .mqtt:  return .mqtt
        case .redis: return .redis
        default:     return nil
        }
    }
}

/// A single port-forwarding rule.
struct PortForward: Codable, Identifiable, Hashable {
    var id: UUID = UUID()
    var type: ForwardType = .local
    /// What the forwarded local port exposes (drives the "Open …" launchers).
    /// Only meaningful for `.local` forwards. `.none` = a plain forward.
    var category: ForwardCategory = .none
    /// Username for a categorized **MQTT / Redis** service, passed to the CLI
    /// client when the tab launches. Not secret, so it lives in the profile; the
    /// matching **password** is stored in the Keychain keyed by this forward's id.
    var serviceUsername: String = ""
    /// Optional bind address for the listening side (e.g. 127.0.0.1, 0.0.0.0, *). Empty = ssh default.
    var bindAddress: String = ""
    /// The port that is opened / listened on (local side for -L and -D, remote side for -R).
    var listenPort: String = ""
    /// The destination host (used by -L and -R, ignored by -D).
    var targetHost: String = "localhost"
    /// The destination port (used by -L and -R, ignored by -D).
    var targetPort: String = ""

    /// Short one-line description for list rows.
    var summary: String {
        let lp = listenPort.isEmpty ? "?" : listenPort
        switch type {
        case .dynamic:
            return "SOCKS :\(lp)"
        case .local:
            return ":\(lp) → \(targetHost):\(targetPort.isEmpty ? "?" : targetPort)"
        case .remote:
            return "srv:\(lp) → \(targetHost):\(targetPort.isEmpty ? "?" : targetPort)"
        }
    }

    /// The address a local client uses to reach this forward. Only `.local`
    /// forwards listen on this Mac, so only they have a local endpoint
    /// (`127.0.0.1:listenPort`, or the chosen bind address).
    var localEndpoint: (host: String, port: Int)? {
        guard type == .local,
              let p = Int(listenPort.trimmingCharacters(in: .whitespaces)), p > 0 else { return nil }
        let bind = bindAddress.trimmingCharacters(in: .whitespaces)
        let host = (bind.isEmpty || bind == "*" || bind == "0.0.0.0") ? "127.0.0.1" : bind
        return (host, p)
    }
}

// A hand-written decoder so profiles saved before `category` existed still load
// (the synthesized one would throw `keyNotFound` on the missing key and, via the
// app's `try?` load, silently drop every saved profile).
extension PortForward {
    enum CodingKeys: String, CodingKey {
        case id, type, category, serviceUsername, bindAddress, listenPort, targetHost, targetPort
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decodeIfPresent(UUID.self, forKey: .id) ?? UUID()
        type = try c.decodeIfPresent(ForwardType.self, forKey: .type) ?? .local
        category = try c.decodeIfPresent(ForwardCategory.self, forKey: .category) ?? .none
        serviceUsername = try c.decodeIfPresent(String.self, forKey: .serviceUsername) ?? ""
        bindAddress = try c.decodeIfPresent(String.self, forKey: .bindAddress) ?? ""
        listenPort = try c.decodeIfPresent(String.self, forKey: .listenPort) ?? ""
        targetHost = try c.decodeIfPresent(String.self, forKey: .targetHost) ?? "localhost"
        targetPort = try c.decodeIfPresent(String.self, forKey: .targetPort) ?? ""
    }
}

/// A reusable, named command the user can insert into a session's terminal.
struct CommandSnippet: Codable, Identifiable, Hashable {
    var id: UUID = UUID()
    var label: String = ""
    var command: String = ""
}

/// A saved web link the user can open in an in-app browser tab. Handy for the
/// web UI a tunnel exposes (e.g. `localhost:8080`) or any related page.
struct ProfileLink: Codable, Identifiable, Hashable {
    var id: UUID = UUID()
    var label: String = ""
    var url: String = ""

    /// The label to show, falling back to the URL when no label is set.
    var displayLabel: String {
        let l = label.trimmingCharacters(in: .whitespaces)
        return l.isEmpty ? url.trimmingCharacters(in: .whitespaces) : l
    }

    /// A best-effort `URL`, adding a scheme when the user omitted one. Local
    /// hosts / bare IPs default to `http://` (typical for tunneled dev UIs);
    /// everything else defaults to `https://`.
    var normalizedURL: URL? {
        var s = url.trimmingCharacters(in: .whitespaces)
        guard !s.isEmpty else { return nil }
        if !s.contains("://") {
            let lower = s.lowercased()
            let looksLocal = lower.hasPrefix("localhost")
                || lower.range(of: #"^\d{1,3}(\.\d{1,3}){3}"#, options: .regularExpression) != nil
            s = (looksLocal ? "http://" : "https://") + s
        }
        return URL(string: s)
    }
}

/// A saved SSH connection + tunnel configuration.
struct SSHProfile: Codable, Identifiable, Hashable {
    var id: UUID = UUID()
    var name: String = "New Profile"
    /// An SF Symbol name chosen for this profile (empty = a sensible default
    /// based on whether it's local). See `displayIcon` and `ProfileIcon`.
    var icon: String = ""
    /// When true this is a **local shell** profile (no SSH): it opens the login
    /// shell in a new tab, starting in `startPath`. All SSH-related fields below
    /// are ignored.
    var isLocal: Bool = false
    /// For local profiles: the folder the shell starts in (supports `~`). Empty
    /// means the default working directory (home).
    var startPath: String = ""
    var host: String = ""
    var port: String = "22"
    var username: String = ""
    /// Path to a private key file (optional). Supports `~` expansion.
    var identityFile: String = ""
    var forwards: [PortForward] = []
    /// When true, open an interactive shell in addition to the tunnels.
    /// When false, pass `-N` (no remote command; tunnels only).
    var openShell: Bool = true
    var compression: Bool = false
    var keepAlive: Bool = true
    var verbose: Bool = false
    /// Optional ProxyJump host (`-J`), e.g. `user@bastion`.
    var jumpHost: String = ""
    /// Extra raw ssh options appended verbatim, e.g. `-o StrictHostKeyChecking=accept-new`.
    var extraOptions: String = ""
    /// The terminal color theme id (see `TerminalTheme`).
    var theme: String = TerminalTheme.defaultID
    /// The terminal text size in points (adjustable live with ⌘+ / ⌘−).
    var fontSize: Double = TerminalFontMetrics.default
    /// Commonly used commands the user can insert into the session's terminal.
    var snippets: [CommandSnippet] = []
    /// Saved web links openable in an in-app browser tab.
    var links: [ProfileLink] = []
    /// Require Touch ID / login password before using the Keychain-stored password.
    var requireAuthForSavedPassword: Bool = true
    /// Optional name of the workspace this profile's tabs should open in.
    /// Connecting switches to (or creates) that workspace. Empty = use whatever
    /// workspace is current.
    var workspace: String = ""

    /// `user@host` style subtitle for list rows.
    var subtitle: String {
        let user = username.isEmpty ? "" : "\(username)@"
        let h = host.isEmpty ? "—" : host
        let p = (port.isEmpty || port == "22") ? "" : ":\(port)"
        return "\(user)\(h)\(p)"
    }

    /// Subtitle shown in the sidebar / palette — local profiles show their start
    /// folder instead of a host.
    var rowSubtitle: String {
        guard isLocal else { return subtitle }
        let p = startPath.trimmingCharacters(in: .whitespaces)
        if p.isEmpty { return "Local shell" }
        let home = NSHomeDirectory()
        let shown = p.hasPrefix(home) ? "~" + p.dropFirst(home.count) : p[...]
        return "Local · \(shown)"
    }

    /// The SF Symbol shown for this profile in the sidebar, editor and palette.
    /// Falls back to a sensible default when the user hasn't picked one.
    var displayIcon: String {
        let chosen = icon.trimmingCharacters(in: .whitespaces)
        if !chosen.isEmpty { return chosen }
        return isLocal ? "terminal" : "point.3.connected.trianglepath.dotted"
    }

    /// The local endpoint of this profile's first dynamic (SOCKS / `-D`) forward,
    /// if any. A web tab opened from this profile routes its traffic through it.
    var socksProxy: (host: String, port: Int)? {
        guard let dyn = forwards.first(where: { $0.type == .dynamic }),
              let port = Int(dyn.listenPort.trimmingCharacters(in: .whitespaces)) else { return nil }
        let bind = dyn.bindAddress.trimmingCharacters(in: .whitespaces)
        return (bind.isEmpty ? "127.0.0.1" : bind, port)
    }

    /// Local forwards tagged with a launchable service category (Web / MQTT /
    /// Redis) that have a usable local endpoint, in profile order. Drives the
    /// "Open …" service launchers in the sidebar and tab menus.
    var categorizedForwards: [PortForward] {
        forwards.filter { $0.category.isLaunchable && $0.localEndpoint != nil }
    }
}

// Defining `init(from:)` in an extension keeps the synthesized memberwise
// initializer, while letting us tolerate older JSON that lacks newer keys.
extension SSHProfile {
    enum CodingKeys: String, CodingKey {
        case id, name, host, port, username, identityFile, forwards, openShell
        case compression, keepAlive, verbose, jumpHost, extraOptions, theme, snippets
        case requireAuthForSavedPassword
        case fontSize
        case isLocal, startPath, icon
        case links
        case workspace
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decodeIfPresent(UUID.self, forKey: .id) ?? UUID()
        name = try c.decodeIfPresent(String.self, forKey: .name) ?? "New Profile"
        icon = try c.decodeIfPresent(String.self, forKey: .icon) ?? ""
        isLocal = try c.decodeIfPresent(Bool.self, forKey: .isLocal) ?? false
        startPath = try c.decodeIfPresent(String.self, forKey: .startPath) ?? ""
        host = try c.decodeIfPresent(String.self, forKey: .host) ?? ""
        port = try c.decodeIfPresent(String.self, forKey: .port) ?? "22"
        username = try c.decodeIfPresent(String.self, forKey: .username) ?? ""
        identityFile = try c.decodeIfPresent(String.self, forKey: .identityFile) ?? ""
        forwards = try c.decodeIfPresent([PortForward].self, forKey: .forwards) ?? []
        openShell = try c.decodeIfPresent(Bool.self, forKey: .openShell) ?? true
        compression = try c.decodeIfPresent(Bool.self, forKey: .compression) ?? false
        keepAlive = try c.decodeIfPresent(Bool.self, forKey: .keepAlive) ?? true
        verbose = try c.decodeIfPresent(Bool.self, forKey: .verbose) ?? false
        jumpHost = try c.decodeIfPresent(String.self, forKey: .jumpHost) ?? ""
        extraOptions = try c.decodeIfPresent(String.self, forKey: .extraOptions) ?? ""
        theme = try c.decodeIfPresent(String.self, forKey: .theme) ?? TerminalTheme.defaultID
        snippets = try c.decodeIfPresent([CommandSnippet].self, forKey: .snippets) ?? []
        links = try c.decodeIfPresent([ProfileLink].self, forKey: .links) ?? []
        requireAuthForSavedPassword = try c.decodeIfPresent(Bool.self, forKey: .requireAuthForSavedPassword) ?? true
        fontSize = TerminalFontMetrics.clamp(try c.decodeIfPresent(Double.self, forKey: .fontSize) ?? TerminalFontMetrics.default)
        workspace = try c.decodeIfPresent(String.self, forKey: .workspace) ?? ""
    }
}

/// Curated SF Symbols offered by the profile icon picker, grouped for display.
enum ProfileIcon {
    static let groups: [(name: String, symbols: [String])] = [
        ("Servers & Network", [
            "point.3.connected.trianglepath.dotted", "server.rack", "network",
            "externaldrive.connected.to.line.below", "cloud", "globe",
            "antenna.radiowaves.left.and.right", "wifi",
        ]),
        ("Devices", [
            "desktopcomputer", "laptopcomputer", "pc", "display",
            "terminal", "macpro.gen3", "tv", "gamecontroller",
        ]),
        ("Storage & Data", [
            "internaldrive", "externaldrive", "cylinder.split.1x2", "tray.full",
            "folder", "shippingbox", "archivebox", "tablecells",
        ]),
        ("Security", [
            "lock.shield", "key", "key.fill", "checkmark.shield", "lock", "hand.raised",
        ]),
        ("Tags & Symbols", [
            "star", "bolt", "flame", "leaf", "heart", "flag",
            "tag", "bookmark", "house", "building.2", "person.crop.circle",
            "gearshape", "hammer", "wrench.and.screwdriver", "ladybug", "cube",
        ]),
    ]

    /// Every symbol, flattened (used for validation / tests).
    static let allSymbols: [String] = groups.flatMap(\.symbols)
}

File: cross-platform/src/RemoteStuff/Models/SshProfile.cs
using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace RemoteStuff.Models;

/// <summary>Shared limits for the terminal text size (points).</summary>
public static class TerminalFontMetrics
{
    public const double Default = 13;
    public const double Min = 8;
    public const double Max = 36;
    public const double Step = 1;

    public static double Clamp(double size) => Math.Min(Max, Math.Max(Min, Math.Round(size)));
}

/// <summary>A single port-forwarding rule.</summary>
public sealed class PortForward
{
    public Guid Id { get; set; } = Guid.NewGuid();

    /// <summary>Optional user-facing name so several forwards can be told apart.</summary>
    public string Name { get; set; } = "";

    public ForwardType Type { get; set; } = ForwardType.Local;

    /// <summary>What the forwarded local port exposes (drives the "Open …" launchers).</summary>
    public ForwardCategory Category { get; set; } = ForwardCategory.None;

    /// <summary>Username for a categorized MQTT / Redis service.</summary>
    public string ServiceUsername { get; set; } = "";

    /// <summary>
    /// Service password captured in the editor, pending persistence into the
    /// encrypted <c>SecretStore</c> (keyed by this forward's <see cref="Id"/>).
    /// Never serialized. Null means "unchanged".
    /// </summary>
    [JsonIgnore]
    public string? PendingServicePassword { get; set; }

    /// <summary>Optional bind address for the listening side (e.g. 127.0.0.1, 0.0.0.0, *).</summary>
    public string BindAddress { get; set; } = "";

    /// <summary>The port that is opened / listened on.</summary>
    public string ListenPort { get; set; } = "";

    /// <summary>The destination host (used by -L and -R, ignored by -D).</summary>
    public string TargetHost { get; set; } = "localhost";

    /// <summary>The destination port (used by -L and -R, ignored by -D).</summary>
    public string TargetPort { get; set; } = "";

    /// <summary>Short one-line description for list rows.</summary>
    [JsonIgnore]
    public string Summary
    {
        get
        {
            var lp = string.IsNullOrEmpty(ListenPort) ? "?" : ListenPort;
            var tp = string.IsNullOrEmpty(TargetPort) ? "?" : TargetPort;
            return Type switch
            {
                ForwardType.Dynamic => $"SOCKS :{lp}",
                ForwardType.Local => $":{lp} → {TargetHost}:{tp}",
                ForwardType.Remote => $"srv:{lp} → {TargetHost}:{tp}",
                _ => ""
            };
        }
    }

    public PortForward Clone() => new()
    {
        Id = Guid.NewGuid(),
        Name = Name,
        Type = Type,
        Category = Category,
        ServiceUsername = ServiceUsername,
        BindAddress = BindAddress,
        ListenPort = ListenPort,
        TargetHost = TargetHost,
        TargetPort = TargetPort
    };

    /// <summary>True when this forward exposes a launchable service (web / MQTT / Redis).</summary>
    [JsonIgnore]
    public bool IsLaunchable => Category.IsLaunchable();

    /// <summary>The label for the "Open …" launcher button.</summary>
    [JsonIgnore]
    public string LaunchLabel => Category switch
    {
        ForwardCategory.Webpage => "Open Web",
        ForwardCategory.Mqtt => "Open MQTT",
        ForwardCategory.Redis => "Open Redis",
        _ => "Open"
    };
}

/// <summary>A reusable, named command the user can insert into a session's terminal.</summary>
public sealed class CommandSnippet
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public string Label { get; set; } = "";
    public string Command { get; set; } = "";
}

/// <summary>One environment variable sent to the server for a session (ssh SetEnv).</summary>
public sealed class EnvVar
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public string Name { get; set; } = "";
    public string Value { get; set; } = "";

    /// <summary>A trimmed <c>NAME=VALUE</c> token, or null when the name is blank.</summary>
    [JsonIgnore]
    public string? SetEnvToken
    {
        get
        {
            var n = Name.Trim();
            return string.IsNullOrEmpty(n) ? null : $"{n}={Value}";
        }
    }
}

/// <summary>A saved web link the user can open in a browser.</summary>
public sealed class ProfileLink
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public string Label { get; set; } = "";
    public string Url { get; set; } = "";
}

/// <summary>A saved SSH connection + tunnel configuration.</summary>
public sealed class SshProfile
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public string Name { get; set; } = "New Profile";

    /// <summary>When true this is a local shell profile (no SSH).</summary>
    public bool IsLocal { get; set; }

    /// <summary>For local profiles: the folder the shell starts in (supports ~).</summary>
    public string StartPath { get; set; } = "";

    public string Host { get; set; } = "";
    public string Port { get; set; } = "22";
    public string Username { get; set; } = "";

    /// <summary>Path to a private key file (optional). Supports ~ expansion.</summary>
    public string IdentityFile { get; set; } = "";

    public List<PortForward> Forwards { get; set; } = new();

    /// <summary>When true, open an interactive shell in addition to the tunnels.</summary>
    public bool OpenShell { get; set; } = true;

    public bool Compression { get; set; }
    public bool KeepAlive { get; set; } = true;
    public bool Verbose { get; set; }

    /// <summary>Optional ProxyJump host (-J), e.g. <c>user@bastion</c>.</summary>
    public string JumpHost { get; set; } = "";

    /// <summary>Extra raw ssh options appended verbatim.</summary>
    public string ExtraOptions { get; set; } = "";

    public bool ForwardAgent { get; set; }
    public bool AddKeysToAgent { get; set; }
    public bool RequestTty { get; set; }

    /// <summary>Seconds before giving up on establishing the connection. 0 = ssh default.</summary>
    public int ConnectTimeout { get; set; }

    public StrictHostKeyChecking StrictHostKeyChecking { get; set; } = StrictHostKeyChecking.Ask;

    /// <summary>A command to run on the server instead of an interactive shell.</summary>
    public string RemoteCommand { get; set; } = "";

    public List<EnvVar> Environment { get; set; } = new();

    /// <summary>A command automatically run in the terminal once the shell is ready.</summary>
    public string RunOnConnect { get; set; } = "";

    public bool AutoConnectOnLaunch { get; set; }
    public bool IsFavorite { get; set; }

    /// <summary>Optional group/folder name for organising the sidebar.</summary>
    public string Group { get; set; } = "";

    public string Theme { get; set; } = "pro";
    public double FontSize { get; set; } = TerminalFontMetrics.Default;

    public List<CommandSnippet> Snippets { get; set; } = new();
    public List<ProfileLink> Links { get; set; } = new();

    // --- Extended options (parity with the macOS app) ---

    /// <summary>An emoji/glyph shown as the profile icon in the sidebar and tabs.</summary>
    public string Icon { get; set; } = "";

    /// <summary>True only when <see cref="Icon"/> is a renderable emoji rather than an
    /// SF Symbol identifier (e.g. <c>macpro.gen3</c>) imported from the macOS app.
    /// SF Symbol names are plain ASCII; emoji contain non-ASCII code points.</summary>
    [JsonIgnore]
    public bool IconIsEmoji
    {
        get
        {
            if (string.IsNullOrWhiteSpace(Icon)) return false;
            foreach (var ch in Icon)
                if (ch > 0x7F) return true;
            return false;
        }
    }

    /// <summary>The icon to show in the UI — empty when <see cref="Icon"/> is a non-renderable
    /// SF Symbol name, so those don't appear as stray text next to the profile name.</summary>
    [JsonIgnore]
    public string DisplayIcon => IconIsEmoji ? Icon : "";

    /// <summary>Optional tab accent colour (hex like <c>#4C8BF5</c>, empty = default).</summary>
    public string TabColor { get; set; } = "";

    /// <summary>Use mosh (mobile shell) instead of plain ssh for the interactive shell.</summary>
    public bool UseMosh { get; set; }

    /// <summary>Automatically reconnect the session if it drops.</summary>
    public bool AutoReconnect { get; set; }

    /// <summary>Log the session's terminal output to a file under the app's Logs folder.</summary>
    public bool LogSession { get; set; }

    /// <summary>How the profile launches: into the current tab area or its own workspace.</summary>
    public WorkspaceLaunch WorkspaceLaunch { get; set; } = WorkspaceLaunch.Current;

    /// <summary>Optional display name for the workspace this profile launches into.
    /// When blank, the profile's own name is used, so several profiles that recreate
    /// the same template each get a distinct, recognisable workspace name.</summary>
    public string WorkspaceName { get; set; } = "";

    /// <summary>The saved-workspace template this profile recreates on launch (by name).
    /// Kept separate from <see cref="WorkspaceName"/> so the workspace can be named after
    /// the profile while still recreating a shared template.</summary>
    public string WorkspaceTemplateName { get; set; } = "";

    /// <summary>True when this profile was created via "Save Workspace as Profile".
    /// Such a launcher re-points its saved workspace template's ad-hoc tabs at this
    /// profile's own host, so relaunching moves the whole workspace to that server.</summary>
    public bool IsWorkspaceLauncher { get; set; }
    // --- Derived / display helpers (not serialized) ---

    /// <summary>
    /// Password captured in the editor, pending persistence into the encrypted
    /// <c>SecretStore</c>. Never serialized to the profiles JSON. Null means "unchanged".
    /// </summary>
    [JsonIgnore]
    public string? PendingPassword { get; set; }

    /// <summary><c>user@host</c> style subtitle for list rows.</summary>
    [JsonIgnore]
    public string Subtitle
    {
        get
        {
            var user = string.IsNullOrEmpty(Username) ? "" : $"{Username}@";
            var h = string.IsNullOrEmpty(Host) ? "—" : Host;
            var p = (string.IsNullOrEmpty(Port) || Port == "22") ? "" : $":{Port}";
            return $"{user}{h}{p}";
        }
    }

    /// <summary>Subtitle shown in the sidebar — local profiles show their start folder.</summary>
    [JsonIgnore]
    public string RowSubtitle
    {
        get
        {
            if (!IsLocal) return Subtitle;
            var p = StartPath.Trim();
            if (string.IsNullOrEmpty(p)) return "Local shell";
            var home = System.Environment.GetFolderPath(System.Environment.SpecialFolder.UserProfile);
            var shown = p.StartsWith(home, StringComparison.Ordinal) ? "~" + p[home.Length..] : p;
            return $"Local · {shown}";
        }
    }

    /// <summary>
    /// The local endpoints (host, port) opened by this profile's <c>-L</c>/<c>-D</c>
    /// forwards, used by the tunnel-health probe to check the ports are listening.
    /// </summary>
    [JsonIgnore]
    public System.Collections.Generic.IEnumerable<(string Host, int Port)> LocalForwardEndpoints
    {
        get
        {
            foreach (var f in Forwards)
            {
                if (f.Type != ForwardType.Local && f.Type != ForwardType.Dynamic) continue;
                if (!int.TryParse(f.ListenPort.Trim(), out var port) || port <= 0) continue;
                var host = f.BindAddress.Trim();
                if (host.Length == 0 || host == "*" || host == "0.0.0.0") host = "127.0.0.1";
                yield return (host, port);
            }
        }
    }

    public SshProfile Clone()
    {
        var copy = (SshProfile)MemberwiseClone();
        copy.Forwards = Forwards.ConvertAll(f => new PortForward
        {
            Id = f.Id,
            Name = f.Name,
            Type = f.Type,
            Category = f.Category,
            ServiceUsername = f.ServiceUsername,
            BindAddress = f.BindAddress,
            ListenPort = f.ListenPort,
            TargetHost = f.TargetHost,
            TargetPort = f.TargetPort
        });
        copy.Environment = Environment.ConvertAll(e => new EnvVar { Id = e.Id, Name = e.Name, Value = e.Value });
        copy.Snippets = Snippets.ConvertAll(s => new CommandSnippet { Id = s.Id, Label = s.Label, Command = s.Command });
        copy.Links = Links.ConvertAll(l => new ProfileLink { Id = l.Id, Label = l.Label, Url = l.Url });
        return copy;
    }
}
